### PR TITLE
Jpegdecoder enhancements

### PR DESCRIPTION
**Pull request by jepler**
* Separate open() [set data source and read metadata] and decode() [actually fill bitmap with pixels] steps
 * add support for any binary stream type. (this unfortuntaely does not enable use of socketpool sockets, see comment below)
   * testing only on io.BytesIO though
 * fix type checking of stream types, this was lost in a micropython merge and not hit by testing
   * (any object with some protocol would ALSO behave like it had a read/write/ioctl protocol, which was wrong)

Note: the testing image is a VERY artifacty blinka jpeg and the output is a simple threshold of the green channel so don't expect a nice portrait.
blinka-q24

### CODE REVIEW

**Comment by jepler** [comment]
The updated test should cover all the related functionality:
 * reading from in-memory data or from an arbitrary object supporting the stream `read` protocol (disk files)
 * x,y offset into destination bitmap
 * x1,y1 - x2,y2 cropping of decoded jpeg data
 * color keying with skip_source_color, which works imperfectly due to the lossy nature of JPEG encoding

Note that this does not enable reading from a socketpool Socket object, because it does not appear to implement the stream `read` protocol, only ioctl:
```c
STATIC const mp_stream_p_t socket_stream_p = {                                               
    .ioctl = socket_ioctl,                                                                   
    .is_text = false,                                                                        
};                                                                                           
```

**Comment by jepler** [comment]
This is now able to stream-decode a JPEG. However, in my test it was not particularly fast.
```
  +0.0000: connecting to wifi
  +3.0908: fetching image
 +16.0797: status 200
 +17.0609: Image size is 196x200
 +17.0610: Decoding image
 +19.0146: Decoding complete
```

<details>
<summary>CircuitPython code for demo on Memento / PyCamera</summary>

```py
import io
import os
import ssl
import time

import displayio
import adafruit_requests
import board
import jpegio
import socketpool
import wifi

class ResponseStreamAdapter(io.IOBase):
    def __init__(self, response):
        self._response = response

    def readinto(self, buf):
        result = self._response._readinto(buf)
        return result

t0 = time.monotonic_ns()
def tprint(msg):
    t1 = time.monotonic_ns()
    ms = (t1 - t0) // 1_000_000
    s = ms // 1000
    ms = ms % 1000
    print(f"{s:+4d}.{ms:04d}: {msg}")

tprint("connecting to wifi")
wifi.radio.connect(os.getenv("WIFI_SSID"), os.getenv("WIFI_PASSWORD"))

pool = socketpool.SocketPool(wifi.radio)
requests = adafruit_requests.Session(pool, ssl.create_default_context())

image_url = "https://timeguy.com/images/avatar.jpg"

j = jpegio.JpegDecoder()
tprint("fetching image")
r = requests.get(image_url)
tprint(f"status {r.status_code}")
if r.status_code == 200:
    stream = ResponseStreamAdapter(r)
    w, h = j.open(stream)
    tprint(f"Image size is {w}x{h}")

    tprint("Decoding image")
    b = displayio.Bitmap(w, h, 65535)
    j.decode(b)
    tprint("Decoding complete")

    gr = displayio.Group()
    tg = displayio.TileGrid(
        b,
        pixel_shader=displayio.ColorConverter(
            input_colorspace=displayio.Colorspace.RGB565_SWAPPED
        ),
    )
    gr.append(tg)
    board.DISPLAY.root_group = gr

    while True:
        pass
else:
    print(r.status_code)
```

</details>

PXL_20231222_161008040

**Comment by jepler** [comment]
This fails on some jpeg images, such as https://cdn-shop.adafruit.com/970x728/5843-05.jpg with "RuntimeError: Not supported JPEG standard"

File: py/stream.c [hunk 1]
```diff
@@ -83,14 +83,13 @@ mp_uint_t mp_stream_rw(mp_obj_t stream, void *buf_, mp_uint_t size, int *errcode
...
 const mp_stream_p_t *mp_get_stream_raise(mp_obj_t self_in, int flags) {
-    const mp_obj_type_t *type = mp_obj_get_type(self_in);
-    if (MP_OBJ_TYPE_HAS_SLOT(type, protocol)) {
-        const mp_stream_p_t *stream_p = MP_OBJ_TYPE_GET_SLOT(type, protocol);
-        if (!((flags & MP_STREAM_OP_READ) && stream_p->read == NULL)
-            && !((flags & MP_STREAM_OP_WRITE) && stream_p->write == NULL)
-            && !((flags & MP_STREAM_OP_IOCTL) && stream_p->ioctl == NULL)) {
-            return stream_p;
-        }
+    // CIRCUITPYTHON-CHANGE: using type-safe protocol accessor
```
Comment (dhalbert) [inline py/stream.c:86]:
```suggestion
    // CIRCUITPY-CHANGE: using type-safe protocol accessor
```

File: py/stream.h [hunk 2]
```diff
@@ -103,7 +103,8 @@ MP_DECLARE_CONST_FUN_OBJ_VAR_BETWEEN(mp_stream_ioctl_obj);
 
 // Object is assumed to have a non-NULL stream protocol with valid r/w/ioctl methods
 static inline const mp_stream_p_t *mp_get_stream(mp_const_obj_t self) {
-    return (const mp_stream_p_t *)MP_OBJ_TYPE_GET_SLOT(((const mp_obj_base_t *)MP_OBJ_TO_PTR(self))->type, protocol);
+    // CIRCUITPYTHON-CHANGE: using type-safe protocol accessor
```
Comment (dhalbert) [inline py/stream.h:106]:
```suggestion
    // CIRCUITPY-CHANGE: using type-safe protocol accessor
```

File: py/makeqstrdata.py [hunk 3]
```diff
@@ -250,6 +250,73 @@
     "zip",
 ]
 
+# These qstrs have to be sorted early (preferably right after static_qstr_list) because they are required to fit in 8-bit values
```
Comment (dhalbert) [inline py/makeqstrdata.py:253]: This is new right, so could you mark it with `CIRCUITPY-CHANGE`?

File: shared-bindings/bitmaptools/__init__.h [hunk 4]
```diff
@@ -88,4 +89,21 @@ void common_hal_bitmaptools_dither(displayio_bitmap_t *dest_bitmap, displayio_bi
...
+typedef struct {
+    union {
+        struct {
+            int16_t x1, y1, x2, y2;
+        };
+        int16_t arr[4];
+    };
+} bitmaptools_rect_t;
+
+#define ARGS_COORD_PAIR ARG_x1, ARG_x2, ARG_x3, ARG_x4
```
Comment (dhalbert) [inline shared-bindings/bitmaptools/__init__.h:101]: Did you mean?
```suggestion
#define ARGS_COORD_PAIR ARG_x1, ARG_y1, ARG_x2, ARG_y2
```
I think I would prefer `ARGS_COORD_PAIR` to be `ARGS_X1_Y1_X2_Y2` or similar so it is more self-evident.

**Review by dhalbert — COMMENTED** [review]
Some touchups. Thanks for the cleanup.

If I had noticed in the earlier PR's, I would have suggested `JPEGDecoder` instead of `JpegDecoder`, but it's late for that now, and it doesn't matter a lot.

**Comment by dhalbert** [comment]
I think this _may_ need a rebase for circuitpython.pot if my 8.2.x update gets merged first

File: py/makeqstrdata.py (hunk 3 above)
Comment (jepler) [reply py/makeqstrdata.py:253]: happily, micropython took a version of this idea so I should go and cherry-pick theirs.

**Comment by jepler** [comment]
I'll likely get back to this sometime next week, thanks for the review!

**Comment by jepler** [comment]
Rebased.

**Review by dhalbert — APPROVED** [review]
Thanks for the updates.

**Pull request merged**